### DISCUSSION TITLE
http3 performance data

### DISCUSSION THREAD

**Discussion by dcl-lily** [curl/curl#15073] [Q&A]
When I request a website using http3 protocol, I find that I can get the index data of time_connect and time_appconnect. What does this data represent? I understand that there is no ssl negotiation process in http3, but it is done in an rrt 
# curl https://bin.bnbstatic.com/static/images/home/register/google-icon.png -s --http3-only  -w "%{time_namelookup}-%{time_connect}-%{time_starttransfer}-%{time_total}-%{speed_download}-%{time_appconnect}-%{time_pretransfer}-%{time_redirect}" -vv 

01:58:11.981103 [0-0] * Host bin.bnbstatic.com:443 was resolved.
01:58:11.981310 [0-0] * IPv6: (none)
01:58:11.981437 [0-0] * IPv4: 3.165.102.47, 3.165.102.49, 3.165.102.38, 3.165.102.123
01:58:11.981506 [0-0] * [HTTPS-CONNECT] added
01:58:11.981556 [0-0] * [HTTPS-CONNECT] connect, init
01:58:11.981629 [0-0] *   Trying 3.165.102.47:443...
01:58:11.986008 [0-0] * [HTTP/3] vquic_send(len=1200, gso=1200) -> 0, sent=1200
01:58:11.986103 [0-0] * [HTTPS-CONNECT] connect -> 0, done=0
01:58:11.986198 [0-0] * [HTTPS-CONNECT] adjust_pollset -> 1 socks
01:58:11.989386 [0-0] * [HTTP/3] populate_x509_store, path=/etc/ssl/certs/ca-certificates.crt, blob=0
01:58:12.019037 [0-0] *  CAfile: /etc/ssl/certs/ca-certificates.crt
01:58:12.019153 [0-0] *  CApath: /etc/ssl/certs
01:58:12.019313 [0-0] * [HTTPS-CONNECT] connect -> 0, done=0
01:58:12.019410 [0-0] * [HTTPS-CONNECT] adjust_pollset -> 1 socks
01:58:12.026480 [0-0] * [HTTP/3] max bidi streams now 128, used 0
01:58:12.026582 [0-0] * [HTTP/3] recvd 4 packets with 4662 bytes -> 0
01:58:12.026961 [0-0] * [HTTP/3] vquic_send_tail_split: [1200 gso=1200][1406 gso=1406]
01:58:12.027099 [0-0] * [HTTP/3] vquic_send(len=1200, gso=1200) -> 0, sent=1200
01:58:12.027214 [0-0] * [HTTP/3] vquic_send(len=1406, gso=1406) -> 0, sent=1406
01:58:12.027265 [0-0] * [HTTP/3] handshake complete after 43ms
01:58:12.028333 [0-0] * Server certificate:
01:58:12.028437 [0-0] *  subject: CN=*.bnbstatic.com
01:58:12.028544 [0-0] *  start date: Sep 25 00:00:00 2024 GMT
01:58:12.028648 [0-0] *  expire date: Oct 23 23:59:59 2025 GMT
01:58:12.028753 [0-0] *  subjectAltName: host "bin.bnbstatic.com" matched cert's "*.bnbstatic.com"
01:58:12.028858 [0-0] *  issuer: C=US; O=Amazon; CN=Amazon RSA 2048 M02
01:58:12.028939 [0-0] *  SSL certificate verify ok.
01:58:12.029022 [0-0] *   Certificate level 0: Public key type RSA (2048/112 Bits/secBits), signed using sha256WithRSAEncryption
01:58:12.029106 [0-0] *   Certificate level 1: Public key type RSA (2048/112 Bits/secBits), signed using sha256WithRSAEncryption
01:58:12.029204 [0-0] *   Certificate level 2: Public key type RSA (2048/112 Bits/secBits), signed using sha256WithRSAEncryption
01:58:12.029306 [0-0] * [HTTP/3] peer verified
01:58:12.029409 [0-0] * [HTTP/3] connect -> 0, done=1
01:58:12.029637 [0-0] * [HTTPS-CONNECT] connect+handshake h3: 48ms, 1st data: 7ms
01:58:12.029728 [0-0] * [HTTPS-CONNECT] connect -> 0, done=1
01:58:12.029834 [0-0] * Connected to bin.bnbstatic.com (3.165.102.47) port 443
01:58:12.029934 [0-0] * using HTTP/3
01:58:12.030075 [0-0] * [HTTP/3] [0] OPENED stream for https://bin.bnbstatic.com/static/images/home/register/google-icon.png
01:58:12.030170 [0-0] * [HTTP/3] [0] [:method: GET]
01:58:12.030271 [0-0] * [HTTP/3] [0] [:scheme: https]
01:58:12.030374 [0-0] * [HTTP/3] [0] [:authority: bin.bnbstatic.com]
01:58:12.030477 [0-0] * [HTTP/3] [0] [:path: /static/images/home/register/google-icon.png]
01:58:12.030600 [0-0] * [HTTP/3] [0] [user-agent: curl/8.11.0-DEV]
01:58:12.030700 [0-0] * [HTTP/3] [0] [accept: */*]
01:58:12.030851 [0-0] * [HTTP/3] vquic_send(len=112, gso=112) -> 0, sent=112
01:58:12.031068 [0-0] * [HTTP/3] [0] cf_send(len=126) -> 126, 0
01:58:12.031142 [0-0] > GET /static/images/home/register/google-icon.png HTTP/3
01:58:12.031142 [0-0] > Host: bin.bnbstatic.com
01:58:12.031142 [0-0] > User-Agent: curl/8.11.0-DEV
01:58:12.031142 [0-0] > Accept: */*
01:58:12.031142 [0-0] > 
............
01:58:12.075241 [0-0] * [HTTP/3] [0] quic close(app_error=256) -> 0
01:58:12.075353 [0-0] * [HTTP/3] recvd 2 packets with 1619 bytes -> 0
01:58:12.075455 [0-0] * [HTTP/3] [0] xfer write failed
01:58:12.075572 [0-0] * [HTTP/3] vquic_send_tail_split: [1444 gso=1444][44 gso=44]
01:58:12.075719 [0-0] * [HTTP/3] vquic_send(len=1444, gso=1444) -> 0, sent=1444
01:58:12.075893 [0-0] * [HTTP/3] vquic_send(len=44, gso=44) -> 0, sent=44
01:58:12.075960 [0-0] * [HTTP/3] [0] cf_recv(blen=102400) -> -1, 23
01:58:12.076012 [0-0] * [HTTP/3] [0] easy handle is done
01:58:12.076122 [0-0] * Connection #0 to host bin.bnbstatic.com left intact
0.021886-0.029763-0.109919-0.116396-0-0.065301-0.072007-0.000000

**Comment by icing** [comment]
With `time_connect` for QUIC we measure in the time it took for curl to receive the first byte from the server. `time_appconnect` is the time for the handshake to finish. QUIC still has a TLS handshake to do, just like over TPC.
